### ISSUE TITLE
BinaryCodable and SIMD aren't compatible 

### ISSUE THREAD

**Issue by kupinski** [christophhagen/BinaryCodable#14]
The SIMD package types conform to Codable but there appears to be an issue with using SIMD types and BinaryCodable.  See the following example code that throws an error.

```swift
import XCTest
import BinaryCodable
import simd
@testable import TestBinaryCodable

struct TestStructure: Codable {
    var x: Int
    var y: String
    var z: Double
    var xx: Array<Float>
    let yy: SIMD2<Double>
}

final class TestBinaryCodableTests: XCTestCase {
    func testExample() throws {
        let strct = TestStructure(x: 42, 
                                  y: "Hello There", 
                                  z: 312.3223, 
                                  xx: [312.12, 565.3,5362.423], 
                                  yy: SIMD2<Double>(52.3,2342.23))
        
        let encoder = BinaryEncoder()
        let data = try encoder.encode(strct)
        
        let decoder = BinaryDecoder()
        let newStrct = try decoder.decode(TestStructure.self, from: data)

        print(strct)
        print(newStrct)
    }
}
```

The error thrown is: 
`
error: -[TestBinaryCodableTests.TestBinaryCodableTests testExample] : failed: caught error: "dataCorrupted(Swift.DecodingError.Context(codingPath: [], debugDescription: "Expected vector with exactly 2 elements.", underlyingError: nil))"
`

**Comment by christophhagen** [comment]
TLDR: The current version of `BinaryCodable` is not compatible with `SIMD` types. A future version will use a different binary format and can support SIMD types.

Explanation:
SIMD types decode the binary data by checking if the `count` property of the `UnkeyedDecodingContainer` is equal to `2`, because it needs exactly two elements.
This property is optional, and `BinaryDecoder` always sets it to `nil`, because it doesn't know how many values are encoded in the container until they are accessed.
The assumed logic is to decode values until the `isAtEnd` property produces `true`.
For example, eight bytes in a container might be decoded as eight `Bool` values, or as a single `Double`.

This restriction can't be resolved without changing the binary format.
I'm currently working on a major overhaul of this library, which will provide a different binary representation (breaking change!), and will support all features that `Codable` offers.
The current binary format, adapted from the much more constrained `Protobuf` encoding, has just proven too limited.

**Comment by kupinski** [comment]
That makes sense.  Thanks for the update.  Very much looking forward to future versions.

**Comment by christophhagen** [comment]
Version 3 of `BinaryCodable` now supports `SIMD` types. But beware, this version is not backwards-compatible with version 2.
